### PR TITLE
Adding the Radklim reader with the stream YAML file

### LINKED ISSUE #216: Data loader for RADKLIM data
### Is your feature request related to a problem? Please describe.

The RADKLIM dataset is available in netCDF-format at JSC (see #107) and can easily be downloaded at other sites as well from DWD's Opendata server.
To enable training/fine-tuning of the WeatherGenerator with this dataset, a netCDF-dataloader must be implemented.

### Describe the solution you'd like

The family of data loaders under `src/weathergen/datasets` will be complemented by a netCDF data loader. 
While the focus is on reading RADKLIM data at first sight, the implementation should also consider possibilities to support reading from netCDF datasets in a more generic way (i.e. minimize hard-coded, dataset-specific assumptions).

### Describe alternatives you've considered

The data could be converted to an anemoi-dataset (or to zarr more generally), but this would lead data duplication.

### Additional context

The RADKLIM dataset shoudl be used during the hackathon in August 2025.

### Organisation

JSC

### PR DESCRIPTION
**Pull request by wael-mika**
## Description

This PR introduces a new data reader for the RADKLIM precipitation dataset, implemented using the base class interface. It also adds the corresponding YAML stream configuration under config/streams/streams_radklim/.

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ X ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #216 

## Code Compatibility

-   [ X ] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ X ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ X ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ X ] I have added comments to my code, particularly in hard-to-understand areas

## Additional Notes

The stream configuration assumes the existence of RADKLIM-compatible reference data and normalization files.

The reader has been structured for easy extension to similar NetCDF-based datasets.

Further testing with training on multiple streams configurations and longer time windows is NECESSARY.

### CODE REVIEW

File: src/weathergen/datasets/data_reader_radklim.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+class RadklimKerchunkReader(DataReaderTimestep):
+    """
+    Reader for RADKLIM data accessed via a Kerchunk reference.
+
+    The reader handles temporal subsetting, lazy loading, and normalization.
+    """
+
+    # Channels used for source and target data
+    source_channels: list[str] = ["RR"]
```
Comment (clessig) [inline src/weathergen/datasets/data_reader_radklim.py]: This should be specified in the stream config and not hard coded, see `select_channels()` in data_reader_anemoi.

File: src/weathergen/datasets/data_reader_radklim.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    target_channels: list[str] = ["RR"]
+    geoinfo_channels: list[str] = []
+
+    # Channel indices
+    source_idx: list[int] = [0]
```
Comment (clessig) [inline src/weathergen/datasets/data_reader_radklim.py]: This should be derived from the source_channels and not hard coded.

File: src/weathergen/datasets/data_reader_radklim.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        # Stack into (time, y, x, var) format
+        arr4 = (
+            ds_win.to_array(dim="var")
+            .transpose("time", "y", "x", "var")
+            .values.astype(np.float32, copy=False)
+        )
+        nt, ny, nx, nvars = arr4.shape
+
+        # Validate channel indices
+        if not channels_idx:
```
Comment (clessig) [inline src/weathergen/datasets/data_reader_radklim.py]: This should be tested at the beginning of the function

File: src/weathergen/datasets/data_reader_radklim.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            raise ValueError("channels_idx must be unique")
+
+        # Flatten spatial/temporal dimensions
+        flat_vars = arr4.reshape(-1, nvars)
+        coords = np.tile(self._base_coords, (nt, 1))
+        time_vals = ds_win["time"].values.astype("datetime64[ns]")
+        times = np.repeat(time_vals, self.points_per_slice)
+
+        # Apply nan filtering
+        valid = ~np.any(np.isnan(flat_vars[:, channels_idx]), axis=1)
```
Comment (clessig) [inline src/weathergen/datasets/data_reader_radklim.py]: No NaNs should be filtered here. This happens later.

File: src/weathergen/datasets/data_reader_radklim.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                num_data_fields=len(channels_idx),
+                num_geo_fields=len(self.geoinfo_idx),
+            )
+
+        # Slice window
+        start, stop = int(t_idxs_rel[0]), int(t_idxs_rel[-1]) + 1
+        ds_win = self.ds.isel(time=slice(start, stop))
+
+        # Stack into (time, y, x, var) format
+        arr4 = (
```
Comment (clessig) [inline src/weathergen/datasets/data_reader_radklim.py]: Could you explain what arr4 is and how it comes out of ds_win

**Review by clessig — COMMENTED** [review]
Will try it out but here already some observations from looking at the code.

File: src/weathergen/datasets/data_reader_radklim.py (hunk 1 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: Sure, I will adjust it, but as I mentioned, I need some guidance at the beginning.

File: src/weathergen/datasets/data_reader_radklim.py (hunk 5 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: This is the dataset for the selected time window
ds_win summary:
<xarray.Dataset> Size: 119MB
Dimensions:  (time: 6, y: 1100, x: 900)
Coordinates:
    lat      (time, y, x) float64 48MB ...
    lon      (time, y, x) float64 48MB ...
  * time     (time) datetime64[ns] 48B 2020-01-03T12:50:00 ... 2020-01-03T17:...
  * x        (x) float64 7kB -4.43e+05 -4.42e+05 -4.41e+05 ... 4.55e+05 4.56e+05
  * y        (y) float64 9kB -4.758e+06 -4.757e+06 ... -3.66e+06 -3.659e+06
Data variables:
    RR       (time, y, x) float32 24MB ...

This is the shape of the array when requesting the variable ('var', 'time', 'y', 'x')
arr4 (before transpose):
  dims : ('var', 'time', 'y', 'x')
  shape: (1, 6, 1100, 900)

Transpose to get (time, y, x, var) this was the requested shape from the older version, I can adapt to any shape you might need
arr4 (after transpose and .values):
  type : <class 'numpy.ndarray'>
  shape: (6, 1100, 900, 1)

File: src/weathergen/datasets/data_reader_radklim.py (hunk 4 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: Ok then, I will remove it.

File: src/weathergen/datasets/data_reader_radklim.py (hunk 3 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: it makes sense, I will move it to the correct location.

File: config/streams/streams_radklim/radklim.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+RADKLIM :
+  type : netcdf
+  referece_path : "/p/scratch/weatherai/data/npp-atms-unpacked/temp_radklim/radklim_output_kerchunk/radklim_full_dataset.json"
```
Comment (tjhunter) [inline config/streams/streams_radklim/radklim.yml]: referece_path : typo
also, this will need to change later to make it agnostic to the hpc

File: src/weathergen/datasets/data_reader_radklim.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+            coords=coords,
+            geoinfos=np.zeros((coords.shape[0], 0), dtype=DType),  # RADKLIM has no geoinfo
+            data=flat_data,
+            datetimes=times,
+        )
+        check_reader_data(rdata, dtr)
+        return rdata
+
+
+def _clip_lat(lats: NDArray[np.floating]) -> NDArray[np.float32]:
```
Comment (tjhunter) [inline src/weathergen/datasets/data_reader_radklim.py]: let's reuse the existing functions instead of copying them around

File: src/weathergen/datasets/data_reader_radklim.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+        kerchunk_ref = json.loads(ref_path.read_text())
+        fs = fsspec.filesystem("reference", fo=kerchunk_ref)
+        mapper = fs.get_mapper("")
+
+        # Consolidate metadata (may be a no-op depending on upstream)
+        with contextlib.suppress(Exception):
+            zarr.consolidate_metadata(mapper)
+
+        ds_full = xr.open_dataset(mapper, engine="zarr", consolidated=True)
```
Comment (tjhunter) [inline src/weathergen/datasets/data_reader_radklim.py]: that's it?? it works just like that? I thought we would have to depend on virtualizarr for the reading too, but I am wrong it seems

Comment (clessig) [reply src/weathergen/datasets/data_reader_radklim.py]: That's the kerchunk version; this are the two magic lines:

```
 fs = fsspec.filesystem("reference", fo=kerchunk_ref)
mapper = fs.get_mapper("")
```

File: config/streams/streams_radklim/radklim.yml (hunk 6 above)
Comment (wael-mika) [reply config/streams/streams_radklim/radklim.yml]: Yes, we certainly will do that later, the idea is to try the data reader if it works, we can then make the ultimate design

File: src/weathergen/datasets/data_reader_radklim.py (hunk 7 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: Sure, we can do that as well

File: src/weathergen/datasets/data_reader_radklim.py (hunk 8 above)
Comment (wael-mika) [reply src/weathergen/datasets/data_reader_radklim.py]: Yes, it is this simple after creating the reference file and then using the little tricks of fsspec.

File: src/weathergen/datasets/data_reader_radklim.py (hunk 7 above)
Comment (clessig) [reply src/weathergen/datasets/data_reader_radklim.py]: Careful here: this function will not correctly convert from an arbitrary convention for spherical coords to the one we expect internally. For this reason, I would keep it local.

Wael, did you check that it is required at all?
